sql/src/main/java/io/crate/execution/dml/upsert/TransportShardIndexAction.java

```diff
@@ -0,0 +1,286 @@
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.crate.exceptions.SQLExceptions.userFriendlyCrateExceptionTopOnly;
+
+@Singleton
+public abstract class TransportShardIndexAction<
+    Request extends ShardRequest<Request, Item>,
+    Item extends ShardRequest.Item, Context
+    > extends TransportShardAction<Request, Item> {
+
+    static final int MAX_RETRY_LIMIT = 100_000; // upper bound to prevent unlimited retries on unexpected states
+
+    public abstract Context buildContext(Request request);
+
+    abstract IndexItemResponse processItem(Item item, Context context, IndexShard indexShard) throws Exception;
+
+    @Inject
+    public TransportShardIndexAction(
+        String actionName,
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        SchemaUpdateClient schemaUpdateClient,
+        TasksService tasksService,
+        IndicesService indicesService,
+        ShardStateAction shardStateAction,
+        Writeable.Reader<Request> reader,
+        IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(
+            actionName,
+            transportService,
+            indexNameExpressionResolver,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            reader,
+            schemaUpdateClient
+        );
+        tasksService.addListener(this);
+    }
+
+    @Override
+    protected WritePrimaryResult<Request, ShardResponse> processRequestItems(IndexShard indexShard,
+                                                                                        Request request,
+                                                                                        AtomicBoolean killed) {
+        ShardResponse shardResponse = new ShardResponse(request.returnValues());
+        Context updateContext = buildContext(request);
+
+        Translog.Location translogLocation = null;
+        for (Item item : request.items()) {
+            int location = item.location();
+            if (killed.get()) {
+                // set failure on response and skip all next items.
+                // this way replica operation will be executed, but only items with a valid source (= was processed on primary)
+                // will be processed on the replica
+                shardResponse.failure(new InterruptedException());
+                break;
+            }
+            try {
+                IndexItemResponse indexItemResponse = processItem(item, updateContext, indexShard);
+                if (indexItemResponse != null) {
+                    if (indexItemResponse.translog != null) {
+                        shardResponse.add(location);
+                        translogLocation = indexItemResponse.translog;
```

Comment:
I was thinking simply remove `IdexItemResponse` and work just on `ShardResponses`. Currently the workflow is:
fold(ShardResponses, (sr, IndexItemResponse) -> sr.add)